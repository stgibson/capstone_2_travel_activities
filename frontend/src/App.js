import React, { useState, useEffect } from "react";
import { useHistory } from "react-router-dom";
import { useDispatch } from "react-redux";
import axios from "axios";
import { v4 as uuid } from "uuid";
import Container from "react-bootstrap/Container";
import Alert from "react-bootstrap/Alert";
import NavBar from "./NavBar";
import Routes from "./Routes";
import { setToken, removeToken } from "./actions/token";

/**
 * Main component of the app
 * @returns JSX code for rendering app
 */
function App() {
  const [currToken, setCurrToken] = useState(null);
  const [errors, setErrors] = useState([]);
  const dispatch = useDispatch();
  const history = useHistory();

  /**
   * Creates new user based on data and stores token if no errors
   * @param {Object{string}} data 
   */
  const signup = async data => {
    try {
      const res = await axios.post(
        `${process.env.REACT_APP_API_BASE_URL}/auth/register`,
        data
      );
      const token = res.data.token;
      localStorage.setItem("token", token);
      dispatch(setToken(token));
      setCurrToken(token);
      history.push("/home");
    }
    catch (err) {
      console.error(err.message);
    }
  };

  /**
   * Authenticates user with data and stores token if no errors
   * @param {Object{string}} data 
   */
  const login = async data => {
    try {
      const res = await axios.post(
        `${process.env.REACT_APP_API_BASE_URL}/auth/login`,
        data
      );
      const token = res.data.token;
      localStorage.setItem("token", token);
      dispatch(setToken(token));
      setCurrToken(token);
      history.push("/home");
    }
    catch (err) {
      console.error(err.message);
    }
  };

  /**
   * Removes token from local storage and redux store
   */
  const logout = () => {
    localStorage.removeItem("token");
    dispatch(removeToken());
  };

  /**
   * Gets list of activities from db based on data
   * @param {Object{string}} data 
   * @returns activities or null if error
   */
  const getActivitiesByLocation = async data => {
    try {
      const res = await axios.get(
        `${process.env.REACT_APP_API_BASE_URL}/activities`,
        {
          params: data,
          headers: { "Authorization": `Bearer ${currToken}` }
        }
      );
      return res.data.activities;
    }
    catch (err) {
      setErrors(errors => [err.message]);
      return null;
    }
  };

  /**
   * Gets all info from db for activity with provided id
   * @param {number} id 
   * @returns activity details or null if error
   */
  const getActivityDetails = async id => {
    try {
      const res = await axios.get(
        `${process.env.REACT_APP_API_BASE_URL}/activities/${id}`,
        {
          headers: { "Authorization": `Bearer ${currToken}` }
        }
      );
      return res.data.activity;
    }
    catch (err) {
      setErrors(errors => [...errors, err.message]);
      return null;
    }
  };

  // when first renders, get token from local storage if it exists
  useEffect(() => {
    setErrors([]);
    const token = localStorage.getItem("token");
    if (token) {
      dispatch(setToken(token));
      setCurrToken(token);
    }
  }, [dispatch]);

  return (
    <div>
      <NavBar logout={ logout } />
      <Container>
        {
          errors.map(err => (
            <Alert key={ uuid() } variant="danger">{ err }</Alert>
          ))
        }
        <Routes
          signup={ signup }
          login={ login }
          getActivitiesByLocation={ getActivitiesByLocation }
          getActivityDetails={ getActivityDetails }
        />
      </Container> 
    </div>
  );
}

export default App;
