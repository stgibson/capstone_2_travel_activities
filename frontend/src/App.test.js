import React from "react";
import { MemoryRouter } from "react-router-dom";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import App from "./App";
import rootReducer from "./reducers/root";

const store = createStore(rootReducer);

it("renders without crashing", () => {
  render(
    <Provider store={ store }><MemoryRouter><App /></MemoryRouter></Provider>
  );
});
