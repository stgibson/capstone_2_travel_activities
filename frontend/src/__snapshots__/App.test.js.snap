// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <nav
      class="navbar navbar-expand navbar-light bg-light"
      varian="light"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Travel Activities
      </a>
      <div
        class="ml-auto navbar-nav"
      >
        <a
          class="nav-link"
          data-rb-event-key="/signup"
          href="/signup"
        >
          Sign Up
        </a>
        <a
          class="nav-link"
          data-rb-event-key="/login"
          href="/login"
        >
          Log In
        </a>
      </div>
    </nav>
    <div
      class="container"
    >
      <h2>
        Log In
      </h2>
      <form
        class=""
      >
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="username"
          >
            Username:
          </label>
          <input
            class="form-control"
            id="username"
            type="text"
            value="testuser"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="password"
          >
            Password
          </label>
          <input
            class="form-control"
            id="password"
            type="password"
            value="testing"
          />
        </div>
      </form>
      <button
        class="btn btn-primary"
        type="button"
      >
        Log In
      </button>
      <nav
        class="navbar navbar-expand navbar-light fixed-bottom"
      >
        <small>
          The information on travel activities in this website is provided by the Tours and Activities API by Amadeus for Developers located at https://developers.amadeus.com/self-service/category/destination-content/api-doc/tours-and-activities/api-reference. Furthermore, this site also uses the Geocoding API located at https://developers.google.com/maps/documentation/geocoding/overview
        </small>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;
